ordering_test.go

```diff
@@ -0,0 +1,464 @@
+				groupID:     descriptor.GroupID,
+				edgeNodeID:  descriptor.EdgeNodeID,
+			},
+			payload: &protobuf.Payload{Seq: proto.Uint64(0)},
+		},
+		{
+			topic: topic{
+				messageType: messageTypeNDATA,
+				groupID:     descriptor.GroupID,
+				edgeNodeID:  descriptor.EdgeNodeID,
+			},
+			payload: &protobuf.Payload{Seq: proto.Uint64(1)},
+		},
+	}
+
+	for _, msg := range messages {
+		processor.processMessage(msg)
+	}
+
+	expected := []sparkplugMessage{messages[0], messages[1]}
+	got := next.messagesByEdgeNode[descriptor]
+	if !reflect.DeepEqual(expected, got) {
+		t.Errorf("expected messages for edge node %s/%s to be in order but weren't", descriptor.GroupID, descriptor.EdgeNodeID)
+	}
+}
+
+func TestBuffersOutOfOrderMessages(t *testing.T) {
+	next := newAppendProcessor()
+	processor := newInOrderProcessor(time.Second, next, newMockRebirther())
+
+	descriptor := EdgeNodeDescriptor{
+		GroupID:    "test-group",
+		EdgeNodeID: "test-node",
+	}
+	messages := []sparkplugMessage{
+		{
+			topic: topic{
+				messageType: messageTypeNBIRTH,
+				groupID:     descriptor.GroupID,
+				edgeNodeID:  descriptor.EdgeNodeID,
+			},
+			payload: &protobuf.Payload{Seq: proto.Uint64(0)},
+		},
+		{
+			topic: topic{
+				messageType: messageTypeNDATA,
+				groupID:     descriptor.GroupID,
+				edgeNodeID:  descriptor.EdgeNodeID,
+			},
+			payload: &protobuf.Payload{Seq: proto.Uint64(2)},
+		},
+		{
+			topic: topic{
+				messageType: messageTypeNDATA,
+				groupID:     descriptor.GroupID,
+				edgeNodeID:  descriptor.EdgeNodeID,
+			},
+			payload: &protobuf.Payload{Seq: proto.Uint64(4)},
+		},
+		{
+			topic: topic{
+				messageType: messageTypeNDATA,
+				groupID:     descriptor.GroupID,
+				edgeNodeID:  descriptor.EdgeNodeID,
+			},
+			payload: &protobuf.Payload{Seq: proto.Uint64(1)},
+		},
+	}
+
+	for _, msg := range messages {
+		processor.processMessage(msg)
+	}
+
+	// we have messages with seq number 0, 2 and 4.
+	// The first one if forwarded as soon as it arrives but 2 and 4 are buffered.
+	// When the message with seq == 1 arrives then we can forward, 1 and 2 but not 4
+	// yet because we are still missing 3.
+	expected := []sparkplugMessage{messages[0], messages[3], messages[1]}
+	got := next.messagesByEdgeNode[descriptor]
+	if !reflect.DeepEqual(expected, got) {
+		t.Errorf("expected messages for edge node %s/%s to be in order but weren't", descriptor.GroupID, descriptor.EdgeNodeID)
+	}
+
+	seq3Message := sparkplugMessage{
+		topic: topic{
+			messageType: messageTypeNDATA,
+			groupID:     descriptor.GroupID,
+			edgeNodeID:  descriptor.EdgeNodeID,
+		},
+		payload: &protobuf.Payload{Seq: proto.Uint64(3)},
+	}
+	processor.processMessage(seq3Message)
+
+	// now that we processed the msg with seq == 3 we can finally forward
+	// that one and the one with seq == 4
+	expected = []sparkplugMessage{messages[0], messages[3], messages[1], seq3Message, messages[2]}
+	got = next.messagesByEdgeNode[descriptor]
+	if !reflect.DeepEqual(expected, got) {
```

Comment:
Why do we use reflection rather than assert equals?

Reply:
we are not using `testify` here. The community is very opinionated when it comes to [assertion libraries](https://google.github.io/styleguide/go/decisions.html#assertion-libraries) so for an open source project I think it's better to stick with the std lib + https://github.com/google/go-cmp for more complex comparisons